scripts/validate-schemas-final.js

```diff
@@ -0,0 +1,104 @@
+const fs = require('fs');
+const path = require('path');
+
+
+function validationDraft04(draft, startFileName, endFileName){
+
+   const Ajv = draft === 'draft04' ? require('ajv-draft-04') : require('ajv');
+   const ajv = new Ajv();
+
+   // Specify the path to the 'schemas' directory
+   const directoryPath = './schemas';
+
+
+   const files = fs.readdirSync(directoryPath);
+
+  // Filter files based on start and end file names
+  const filteredFiles = files.filter(file => {
+    return file >= startFileName && file <= endFileName;
+  });
+
+
+    // Iterate through the filtered files
+  filteredFiles.forEach(file => {
+    // Construct the full path to the JSON schema file
+    const filePath = path.join(directoryPath, file);
+
+    try {
+      // Read and parse the JSON schema
+      const fileContent = fs.readFileSync(filePath, 'utf8');
+      const obj = JSON.parse(fileContent);
+
+      // Remove unnecessary definitions
+      delete obj.definitions['http://json-schema.org/draft-04/schema'];
+      delete obj.definitions['http://json-schema.org/draft-07/schema'];
+
+      // Compile the schema
+      const validate = ajv.validateSchema(obj);
+
+      // Check if the schema is valid
+      if (validate) {
+        console.log(`${file}: JSON Schema is valid!`);
+      } else {
+        console.error(`${file}: JSON Schema is not valid:`, ajv.errors);
+        process.exit(1);
+      }
+    } catch (error) {
+      console.error(`${file}: Error reading or parsing JSON Schema:`, error.message);
+      process.exit(1);
+    }
+  });
+}
+
+
+function validation(draft, startFileName, excludedFiles){
+
+    const Ajv = draft === 'draft04' ? require('ajv-draft-04') : require('ajv');
+    const ajv = new Ajv();
+ 
+    // Specify the path to the 'schemas' directory
+    const directoryPath = './schemas';
+ 
+ 
+    const files = fs.readdirSync(directoryPath);
+ 
+   // Filter files based on start and end file names
+   const filteredFiles = files.filter(file => {
+     return file >= startFileName && !excludedFiles.includes(file);
+   });
+ 
+ 
+    // Iterate through the filtered files
+   filteredFiles.forEach(file => {
+     // Construct the full path to the JSON schema file
+     const filePath = path.join(directoryPath, file);
+ 
+     try {
+       // Read and parse the JSON schema
+       const fileContent = fs.readFileSync(filePath, 'utf8');
+       const obj = JSON.parse(fileContent);
+ 
+       // Remove unnecessary definitions
+       delete obj.definitions['http://json-schema.org/draft-04/schema'];
+       delete obj.definitions['http://json-schema.org/draft-07/schema'];
```

Comment:
I believe you said this wasn't necessary when using the `ajv.validateSchema` right? What happens when we remove those two lines?

Reply:
let me try running the script without including these lines

Reply:
I have some observations I would like to report regarding this:

- If I remove the above said lines from the `validationDraft04` function, it gives no error.
- But if I remove that from the `validation` function (which validates the draft-07 files), it gives the following error :point_down: 
```
1.0.0-without-$id.json: JSON Schema is valid!
1.0.0.json: JSON Schema is valid!
1.1.0-without-$id.json: JSON Schema is valid!
1.1.0.json: JSON Schema is valid!
1.2.0-without-$id.json: JSON Schema is valid!
1.2.0.json: JSON Schema is valid!
2.0.0-rc1.json: JSON Schema is not valid: [
  {
    instancePath: '/definitions/http:~1~1json-schema.org~1draft-04~1schema/properties/multipleOf/exclusiveMinimum',
    schemaPath: '#/properties/exclusiveMinimum/type',
    keyword: 'type',
    params: { type: 'number' },
    message: 'must be number'
  }
]
```

If it allows, I think we should keep those lines for the sake of not breaking the overall validation process :sweat_smile: Thoughts? :eyes:

Reply:
I don't really think that error is related with you removing those lines. Are you sure that's the reason?

Reply:
You can use an external JSON Schema validator in order to see if those files are valid or not.
I've test it, and both seem invalid.

- `2.0.0-rc1.json`: https://www.jsonschemavalidator.net/s/r114nPn0
- `2.0.0-rc1-without-$id.json`: https://www.jsonschemavalidator.net/s/gocJb7o3

So maybe we just discovered, thanks to your script, that those have been there in a broken state for a while.
What do you believe we should do then with those two schemas? If they are not valid, does it make sense to even keep them here?

Pinging the rest of code owners here @fmvilas @derberg @dalelane @char0n.

Reply:
Coming back to your script, the weird thing is that the `2.0.0-rc1.json` seems valid in your script, meanwhile `ajv` cli says both are invalid:

```
❯ ajv compile -s './schemas/2.0.0-rc1*'
schema ./schemas/2.0.0-rc1-without-$id.json is invalid
error: schema is invalid: data/definitions/json-schema-draft-07-schema/properties/multipleOf/exclusiveMinimum must be number
schema ./schemas/2.0.0-rc1.json is invalid
error: schema is invalid: data/definitions/http:~1~1json-schema.org~1draft-04~1schema/properties/multipleOf/exclusiveMinimum must be number
```

Reply:
It seems you don't need those:

```
delete obj.definitions['http://json-schema.org/draft-04/schema'];
delete obj.definitions['http://json-schema.org/draft-07/schema'];
```

It seems the way `validateSchema` works is that it really requires the schemas to be present and that It does not add any meta schema by default.

Reply:
> Coming back to your script, the weird thing is that the `2.0.0-rc1.json` seems valid in your script, meanwhile `ajv` cli says both are invalid (expected):
> 
> ```
> ❯ ajv compile -s './schemas/2.0.0-rc1*'
> schema ./schemas/2.0.0-rc1-without-$id.json is invalid
> error: schema is invalid: data/definitions/json-schema-draft-07-schema/properties/multipleOf/exclusiveMinimum must be number
> schema ./schemas/2.0.0-rc1.json is invalid
> error: schema is invalid: data/definitions/http:~1~1json-schema.org~1draft-04~1schema/properties/multipleOf/exclusiveMinimum must be number
> ```

If I delete this line from the script, 
```
delete obj.definitions['http://json-schema.org/draft-04/schema']:
```
then only it says that `2.0.0-rc1.json` is invalid.

Reply:
> It seems you don't need those:
> 
> ```
> delete obj.definitions['http://json-schema.org/draft-04/schema'];
> delete obj.definitions['http://json-schema.org/draft-07/schema'];
> ```
> 
> It seems the way `validateSchema` works is that it really requires the schemas to be present and that It does not add any meta schema by default. See:
> 
> > If schema has $schema property, then the schema with this id (that should be previously added) is used to validate passed schema.
> > Source: https://ajv.js.org/api.html#ajv-validateschema-schema-object-boolean

Okay, I see. I'll remove this.

Does it mean that it is already doing the `delete` statements inclusively that we were dong exclusively in the script?

Reply:
> You can use an external JSON Schema validator in order to see if those files are valid or not. I've test it, and both seem invalid.
> 
>     * `2.0.0-rc1.json`: https://www.jsonschemavalidator.net/s/r114nPn0
> 
>     * `2.0.0-rc1-without-$id.json`: https://www.jsonschemavalidator.net/s/gocJb7o3
> 
> 
> So maybe we just discovered, thanks to your script, that those have been there in a broken state for a while.

It is already being helpful before fully deployed, :eyes:  happy to see that :grinning: 
 
> What do you believe we should do then with those two schemas? If they are not valid, does it make sense to even keep them here?

I think if they are not being implemented anywhere, it would be best to delete them.

I see `rc-1`, `rc-2`, and then there is the `2.0.0.json` with these three having 2 files for each, one with `$id` and another `without-$id`. What is the difference between them? And which one out of those is used for real-world scenarios for v2?

Reply:
Follow up issue was created https://github.com/asyncapi/spec-json-schemas/issues/455